### PR TITLE
📖 Topology Mutation Hook proposal

### LINKED ISSUE #6319: RFE: ClusterClass external patches
Currently, ClusterClass allows customization of clusters only with inline patches. Goal of this issue is to propose  RuntimeHooks which enable users to implement RuntimeExtensions which generate patches externally.

This is a follow-up to the ClusterClass proposal:
> Extensibility/Transformation:
>   Provide an external mechanism which allows to leverage the full power of a programming language to implement more complex customizations.
> https://github.com/kubernetes-sigs/cluster-api/blob/78dd064d410cccf4617b69e21d35b1c00448eb99/docs/proposals/20210526-cluster-class-and-managed-topologies.md#prospective-future-work

This has been discussed a few times in the past in the context of ClusterClass and during the implementation of inline patches. We also have a few open issues which show that users are now hitting the limitations of inline patches (e.g. https://github.com/kubernetes-sigs/cluster-api/issues/6245)

The proposal builds on top of the foundation of the [ClusterClass](https://github.com/kubernetes-sigs/cluster-api/blob/78dd064d410cccf4617b69e21d35b1c00448eb99/docs/proposals/20210526-cluster-class-and-managed-topologies.md) and the [RuntimeSDK](https://github.com/kubernetes-sigs/cluster-api/pull/6181) proposals.

/area topology
/kind design
/kind proposal

### PR DESCRIPTION
**Pull request by sbueringer**
Signed-off-by: Stefan Büringer buringerst@vmware.com

**What this PR does / why we need it**:
Proposal to address #6319

Tasks:
* [x] [After #6181] merge] Update links to Runtime SDK proposal
* [x] [Before merge] Update Swagger UI links to main branch

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #6319

### CODE REVIEW

File: docs/proposals/20220330-topology-mutation-hook.md
Comment (johannesfrey) [inline docs/proposals/20220330-topology-mutation-hook.md:144]: Just for my understanding. This means that an `Extension` must always be deployed before a `ClusterClass` that references it, because the computed name is unkown beforehand and will only be visible in the `status` after the `Extension` has been registered successfully?

Also why is the field named `hook` when it actually references an `extension`? Shouldn't this reference the name field like mentioned here: https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20220221-runtime-SDK.md?plain=1#L559
By the above sentence I would think of using the name of the field here instead: https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20220221-runtime-SDK.md?plain=1#L562
But probably I'm just conflating the terms hook and extension here :sweat_smile:

Comment (fabriziopandini) [reply docs/proposals/20220330-topology-mutation-hook.md:144]: > This means that an Extension must always be deployed before a ClusterClass that references it, because the computed name is unkown beforehand and will only be visible in the status after the Extension has been registered successfully?

AFAIK as I remember computed name value is predictable, it is the concatenation of the name returned by discovery and the name the Extension object 
@sbueringer to confirm

> Also why is the field named hook when it actually references an extension? 

This makes sense to me, +1 to disambiguate if possible

Comment (sbueringer) [reply docs/proposals/20220330-topology-mutation-hook.md:144]: > AFAIK as I remember computed name value is predictable, it is the concatenation of the name returned by discovery and the name the Extension object

That is correct. If you use a well-defined and documented External Patch Extension the name is deterministic. Otherwise deploying concurrently is not an issue as we retry the Cluster reconciliation until the the corresponding `Extension` exists and discovery is done.

> Also why is the field named hook when it actually references an extension?

Agree. The current name is not ideal. I would wait for the upcoming Extension CRD PR though as I think we have to think about how we call "Discovered RuntimeExtensions" there and this might lead to an obvious name here.

**Comment by fabriziopandini** [comment]
@sbueringer In order to gather some feedback let's ping the appointed reviewers (If someone else is interested, we should make sure to add them to this list as well).

**Comment by sbueringer** [comment]
@CecileRobertMichon @enxebre @vincepri @killianmuldoon @ykakarap
If you have some time to review this proposal (if not already done), I would appreciate it :)

File: docs/proposals/20220330-topology-mutation-hook.md
Comment (ykakarap) [reply docs/proposals/20220330-topology-mutation-hook.md:144]: Just FYI: The Extension CRD PR is now merged. Naming the field `extensionHandler` will be in sync with the naming used in the Extension CRD. Open to other suggestions.

**Review by ykakarap — COMMENTED** [review]
The proposal is LGTM pending the naming comment. 

Exciting improvements to ClusterClass 🚀

File: docs/proposals/20220330-topology-mutation-hook.md
Comment (sbueringer) [reply docs/proposals/20220330-topology-mutation-hook.md:144]: I'm not sure if we just call it extensionHandler that the external patching aspect is clear enough.
I see the following variations / dimensions:
1. do we want to the fields top-level or nested (like we did it with inline patches with the "definition" field)
2. how do we name the fields

One idea:
```yaml
  patches:
  - name: external-patch-1
    external:
      generate:
      validate:
  - name: config
    definitions:
    - selector:
    ....
```

Alternatives:
1. inline the fields instead of having them in `external`
2. Alternative field variations:
* generate+validate
* generatePatches+validateTopology
* mutate+validate
* generationHandler+validationHandler
* generatePatchesHandler+validateTopologyHandler
* mutationHandler+validationHandler
* Similar with "Extension": generationExtension+validationExtension, ...


@fabriziopandini @killianmuldoon @vincepri @ykakarap Opinions?

Comment (ykakarap) [reply docs/proposals/20220330-topology-mutation-hook.md:144]: I would vote for having nested fields under `external`. As for the field names how about `generator` (generates the patch) and `validator` (validates the topology)?

Names from the above list that are also okay:
* generatePatches+validateTopology
* generationHandler+validationHandler

Let's not use mutate because it doesn't really mutate the object it only send back the patch to be applied. Using mutate could have a different impression.

Comment (fabriziopandini) [reply docs/proposals/20220330-topology-mutation-hook.md:144]: +1 to nest with generateExtension+validateExtension

**Review by killianmuldoon — APPROVED** [review]
Awesome stuff!

/lgtm

**Review by vincepri — APPROVED** [review]
/approve
/lgtm
/hold

File: docs/proposals/20220330-topology-mutation-hook.md
Comment (sbueringer) [inline docs/proposals/20220330-topology-mutation-hook.md:225]: This link and the one below only works after merge. For review this one points to the file on the PR: https://editor.swagger.io/?url=https://raw.githubusercontent.com/kubernetes-sigs/cluster-api/1c39879fa72a7dd77b8827f6acec2e4cca8e492a/docs/proposals/images/topology-mutation-hook/runtime-sdk-openapi.yaml

**Review by CecileRobertMichon — COMMENTED** [review]
/lgtm

**Comment by fabriziopandini** [comment]
/hold cancel
/lgtm
/approve
Lazy consensus deadline expired

**Pull request merged**
